readProperty and propertyReading messages

benfrancis opened the pull request:
This PR sets out a basic message format and defines the first two message types - `readProperty` and `propertyReading`.

It is intended to fulfill requirements [5.2.2](https://www.w3.org/community/reports/web-thing-protocol/CG-FINAL-web-thing-protocol-requirements-20231101/#websocket-sub-protocol-properties).1 and [5.2.2](https://www.w3.org/community/reports/web-thing-protocol/CG-FINAL-web-thing-protocol-requirements-20231101/#websocket-sub-protocol-properties).11 from the Use Cases & Requirements document.

Some notable differences from the [Strawman Proposal](https://docs.google.com/document/d/1KWv-aQfMgsqBFg0v4rVqzcVvzzisC7y4X4CMUYGc8rE/edit?usp=sharing) include:
- Thing ID capitalised as `thingID` rather than `thingId`
- The addition of `messageID` and `correlationID` members as per the proposals by @RobWin and @hspaay (#35, #31)
- `property` is renamed to `name` as suggested by @hspaay (in #34)
- `data` is renamed to `value` to clearly establish a `name` and `value` pair (The term `data` is only used for the data schema of events in the Thing Description specification)
- I haven't yet added error conditions or the `error` message format, which will come in a future PR

This is an important PR which establishes the basic pattern of what messages look like. That doesn't mean we can't change that basic format in the future if a better solution comes along, but it will at least set our initial direction.

Some open questions I have:
- The term `name` is intended to be re-usable between multiple message types (e.g. for the names of action and event affordances as well), but is it explicit enough? Should it be `affordanceName` or `propertyName`/`actionName`/`eventName` instead for clarity?
- I'm still wondering about @hspaay's proposal (in #34) to split out a separate `operation` member which directly takes the WoT operation name like `readproperty`, then a separate member to distinguish between different message types within in an operation. I'm reluctant to do that because I'm not sure there's an established enough WoT taxonomy for distinguishing between those messages (e.g. "request", "response" and "notification" as @hspaay proposed) and we might be optimising prematurely. My instinct is to just have a single `messageType` member which encapsulates the entire purpose of the message, but I'm still open to other opinions.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/benfrancis/web-thing-protocol/pull/37.html" title="Last updated on Jan 29, 2025, 3:25 PM UTC (547bb95)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/web-thing-protocol/37/92add2c...benfrancis:547bb95.html" title="Last updated on Jan 29, 2025, 3:25 PM UTC (547bb95)">Diff</a>

egekorkan commented on index.html:
```diff
@@ -215,8 +216,204 @@ <h3>WebSocket Re-use</h3>
+            </tr>
+            <tr>
+              <td><code>value</code></td>
+              <td>string</td>
```
Does it mean that values should be stringified first?

egekorkan replied:
ok can be resolved

egekorkan commented on index.html:
```diff
@@ -215,8 +216,204 @@ <h3>WebSocket Re-use</h3>
+              <td><code>timestamp</code></td>
+              <td>string</td>
+              <td>Optional</td>
+              <td>A timestamp in date-time format [[RFC3339]] set to the time the property reading took place.</td>
```
Do we requite UTC or any time zone can be used (allowed by this rfc)

benfrancis replied:
RFC 3339 only allows numerical offsets relative to UTC, and "Z" as timezones. I've added an assertion to the WebSocket Messages section to highlight this, consistent with WoT Profiles.

egekorkan commented on index.html:
```diff
@@ -215,8 +216,204 @@ <h3>WebSocket Re-use</h3>
+            <tr>
+              <td><code>value</code></td>
+              <td>string</td>
+              <td>Mandatory</td>
```
nit: what if there is no value atm? I guess that should be an error case right?

benfrancis replied:
Hmm, so the property exists but doesn't have a value? I guess it could return a value of `null` or `undefined`?

benfrancis replied:
@hspaay wrote:
> Is no value really an error case?

If the request is to read the value of a property and the value of the property couldn't be read, then yes that feels like an error case to me.

> It is possible that the Thing simply hasn't been able to obtain a value yet (undefined in JS). Requiring a value implies that the Thing is required to use a default. Defaults are however optional in the TD.
> Is there an official definition of 'not-a-value' in JSON?

I think that would be `null`. `undefined` would imply to me that there is no property of that name, and anyway `undefined` is not a valid value in JSON. I think a property could theoretically return a value of `null`, but only if the JSON schema explicitly allows a null value, e.g.

```json
"type": ["string", "null"]
```

Otherwise it would fail JSON Schema validation.

Does that make sense..?

--- 

In the Strawman proposal I listed the following error conditions (which still need some refinement):
- If the Thing does not have a Property with the requested name, the Thing MUST respond with a Not Found [error](https://docs.google.com/document/d/1KWv-aQfMgsqBFg0v4rVqzcVvzzisC7y4X4CMUYGc8rE/edit#heading=h.eh17hg7lhhfi) message.
- If the Thing fails to read the requested Property value because the Property is write-only, it SHOULD respond with an Operation Not Allowed [error](https://docs.google.com/document/d/1KWv-aQfMgsqBFg0v4rVqzcVvzzisC7y4X4CMUYGc8rE/edit#heading=h.eh17hg7lhhfi) message.
- If the Thing fails to read the requested Property value for any other reason, it MAY respond with a Thing Error [error](https://docs.google.com/document/d/1KWv-aQfMgsqBFg0v4rVqzcVvzzisC7y4X4CMUYGc8rE/edit#heading=h.eh17hg7lhhfi) message

hspaay replied:
Yes that makes sense. :+1:

benfrancis commented on index.html:
```diff
@@ -215,8 +216,204 @@ <h3>WebSocket Re-use</h3>
   <section id="websocket-messages">
     <h2>WebSocket Messages</h2>
+    <p>All messages MUST be serialised using the JSON [[JSON]] data interchange format.</p>
+    <p>All messages MUST contain a top level JSON <a href="https://www.rfc-editor.org/rfc/rfc8259#section-4">object</a>
```
Hmm, I felt that these were two separate assertions:
1. That it should be a JSON document
2. That the JSON document should always contain a top level object (which not all JSON documents do)

I guess they could be combined into one, but is that as clear?

benfrancis replied:
OK, on further reflection I agree it's enough to just say messages must be a JSON object. I've combined them into one assertion.

benfrancis commented on index.html:
```diff
@@ -215,8 +216,204 @@ <h3>WebSocket Re-use</h3>
+      <p>The following sections define WebSocket message payload formats for reading, writing and observing
+        <a>Properties</a> of <a>Things</a>.
+      </p>
+      <section id="readProperty">
```
Apart from following the JSON conventions used everywhere else in the message payload the different capitalisation was intentionally meant to differentiate it from an operation name, since it isn't just an operation name. It is a message type which incorporates both the type of operation and the part of the operation that the message provides. There are message types which aren't also operation names like `propertyReading`, `actionStatus` and `error` so I thought making them look too much like operation names might get confusing.

However, there is some debate as to whether `operation` should be split out into a separate member which directly takes an operation name from the WoT TD specification. I'd be interested in your view on this. See #34.

egekorkan replied:
ok without #34 proposal on splitting, this capitalisation is indeed better so my initial question is answered. I will comment there

benfrancis commented on index.html:
```diff
@@ -215,8 +216,204 @@ <h3>WebSocket Re-use</h3>
+    <p>The top level JSON object MUST contain a <code>thingID</code> member with the value set to a unique identifier of
+      the <a>Web Thing</a> to which the message relates.
+      If the <a>Thing Description</a> of the <a>Web Thing</a> contains an <code>id</code> member then the value of that
+      <code>id</code> member MUST be used as the unique identifier.
```
I have re-worded this, and the next assertion, for clarity.

egekorkan replied:
Looks good. Can be resolved

benfrancis commented on index.html:
```diff
@@ -215,8 +216,204 @@ <h3>WebSocket Re-use</h3>
+    <p>The top level JSON object MUST contain a <code>messageID</code> member with the value set to a unique identifier
+      for the current message in UUIDv4 format [[rfc9562]].</p>
+    <p>The top level JSON object MAY contain a <code>correlationID</code> member which provides a unique identifier in
+      UUIDv4 format [[rfc9562]] which is shared between messages corresponding to the same WoT operation (e.g. a request
```
Yes, I've added an additional example and this will be elaborated on further in the `subscribeEvent` and `observeProperty` sections.

egekorkan replied:
ok makes sense. This PR is not about those

RobWin commented on index.html:
```diff
@@ -215,8 +216,215 @@ <h3>WebSocket Re-use</h3>
+            "messageType": "propertyReading",
+            "name": "on",
+            "value": true,
+            "timestamp": "2024-01-13T23:20:50.52Z",
```
I think we had a dedicated discussion on this in the Strawman proposal. However, I believe we need to two timestamp in the final version:

* Creation Timestamp: Indicates when the message was created.
* Update Timestamp: (Optional) Reflects when the property was last updated/changed. This can be omitted if the device cannot provide it.

Currently, I understand that the timestamp represents when the property was read (the message was created), but it does not indicate when the property was updated/changed. 
If we add both timestamp, should we then use more clear names for the timestamps?

benfrancis replied:
I see what you mean, kind of like `Date` vs. `Last-Modified` in HTTP response headers.

I'm not sure what we'd call them. Maybe `timestamp` and `lastModifiedTimestamp`...

I'm curious whether any existing protocol bindings include a last modified timestamp in property readings because this is obviously not something that all Things can provide and almost (but not quite) feels like a separate property (e.g. asking a light switch when it was last turned on). @egekorkan @relu91?

RobWin replied:
I definitely think that a lastUpdated or lastChanged timestamps are not separate properties, because you might want to have these timestamps on all properties. Either it's additional metadata of a property or the property would need to be a JSON object.
Otherwise it would lead to an explosion of properties.

hspaay replied:
Wrt lastModified vs lastUpdated, if a property has been 'refreshed' with the same value, I consider this an update but not a modification. (for example in case a device refreshes property values at least once a day to confirm the value is still up to date).

RobWin replied:
Exactly, there is a semantic difference between lastUpdated and lastModified/lastChanged.

benfrancis commented on index.html:
```diff
@@ -215,8 +216,215 @@ <h3>WebSocket Re-use</h3>
+              <td><code>name</code></td>
+              <td>string</td>
+              <td>Mandatory</td>
+              <td>The name of the <a>Property</a> to read.</td>
```
I've added "as per its key in the properties member of the Thing Description" to the description of the `name` member in both messages. Does that make sense?

benfrancis commented:
OK this PR has been open for a couple of weeks now, I think I've addressed the review comments and we have two approvals so I'm going to land it.

I'm still interested in exploring splitting `operation` out as a separate member, but there are some open questions in #34 as to how that would work for some operation types.

Thank you for the feedback everyone!

relu91 commented:
@benfrancis sorry for arriving late, I just want to say that I read the text and I haven't seen any obvious shortcomings. Maybe the messageType of the readproperty operation could be more obvious like `propertyResult` or `propertyValue`, but I can also live with `propertyReading`.
